OrbitProducer/include/OrbitProducer/LockFreeBufferCaptureEventProducer.h

```diff
@@ -0,0 +1,141 @@
+
+#include "OrbitBase/Logging.h"
+#include "OrbitProducer/CaptureEventProducer.h"
+#include "concurrentqueue.h"
+
+namespace orbit_producer {
+
+// This still abstract implementation of CaptureEventProducer provides a lock-free queue
+// where to write events with low overhead from the fast path where they are produced.
+// The methods EnqueueIntermediateEvent(IfCapturing) allow to enqueue those events.
+//
+// Internally, a thread reads from the lock-free queue and sends CaptureEvents
+// to ProducerSideService using the methods provided by the superclass.
+//
+// Note that the events stored in the lock-free queue, whose type is specified by the
+// type parameter IntermediateEventT, don't need to be CaptureEvents, nor protobufs at all.
+// This is to allow enqueuing objects that are faster to produce than protobufs.
+// The translation from IntermediateEventT to CaptureEvent is handled by
+// TranslateIntermediateEvent, which subclasses need to implement.
+template <typename IntermediateEventT>
+class LockFreeBufferCaptureEventProducer : public CaptureEventProducer {
+ public:
+  void BuildAndStart(const std::shared_ptr<grpc::Channel>& channel) override {
+    CaptureEventProducer::BuildAndStart(channel);
+
+    forwarder_thread_ = std::thread{[this] { ForwarderThread(); }};
+  }
+
+  void ShutdownAndWait() override {
+    shutdown_requested_ = true;
+
+    CHECK(forwarder_thread_.joinable());
+    forwarder_thread_.join();
+
+    CaptureEventProducer::ShutdownAndWait();
+  }
+
+  void EnqueueIntermediateEvent(const IntermediateEventT& event) {
+    lock_free_queue_.enqueue(event);
+  }
+
+  void EnqueueIntermediateEvent(IntermediateEventT&& event) {
+    lock_free_queue_.enqueue(std::move(event));
+  }
+
+  bool EnqueueIntermediateEventIfCapturing(
+      const std::function<IntermediateEventT()>& event_builder_if_capturing) {
+    if (IsCapturing()) {
+      lock_free_queue_.enqueue(event_builder_if_capturing());
+      return true;
+    }
+    return false;
+  }
+
+ protected:
+  void OnCaptureStart() override {
+    CaptureEventProducer::OnCaptureStart();
+    {
+      absl::MutexLock lock{&should_send_all_events_sent_mutex_};
+      should_send_all_events_sent_ = false;
+    }
+  }
+
+  void OnCaptureStop() override {
+    CaptureEventProducer::OnCaptureStop();
+    {
+      absl::MutexLock lock{&should_send_all_events_sent_mutex_};
+      should_send_all_events_sent_ = true;
+    }
+  }
+
+  // Subclasses need to implement this method to convert an IntermediateEventT enqueued
+  // in the internal lock-free buffer to a CaptureEvent to be sent to ProducerSideService.
+  [[nodiscard]] virtual orbit_grpc_protos::CaptureEvent TranslateIntermediateEvent(
+      IntermediateEventT&& intermediate_event) = 0;
+
+ private:
+  void ForwarderThread() {
+    constexpr uint64_t kMaxEventsPerRequest = 10'000;
+    std::vector<IntermediateEventT> dequeued_events;
+    dequeued_events.resize(kMaxEventsPerRequest);
```

Comment:
nit: you can do it in c-tor.

Reply:
This is a local variable (and there's no need for it to not be a local variable).

Reply:
replace 86 and 87 with
std::vector<IntermediateEventT> dequeued_events(kMaxEventsPerRequest);

Reply:
Ah ok, sorry, I had obviously read your comment completely wrong...

Reply:
Done.